### ISSUE TITLE
Deprecate HttpServer and primitive REST model in favor of a better solution

### ISSUE THREAD

**Issue by tylerjwatson** [NyxStudios/TShock#940]
### Preface

The REST interface is probably one of the best intergation features of TShock.  It allows people to fetch information never previously possible from remote servers.  It serves as the web integration and administration point for every terraria server there is.

### Problem

The `HttpServer` module used in TShock is extremely slow; requests to it can take up to 500ms using my own tests.  Each request to the `HttpServer` listener has about 6 `NullRefererenceException` objects thrown and swallowed, severely hurting its RTT performance.

The `HttpServer` module was inserted a long time ago in lieu of a workable HTTP server existing inside the .net framework at the time.

The REST object interface is also quite restrictive, and lacks support for modern C# programming techniques such as anonymous types, `dynamic`, and route-based action selectors.

### Recommendation

.It's now possible to host a full Web API 2.0 server inside any .NET app.  The OWIN stack is used by ASP.net MVC, Web API 2.0 and a lot of next-generation technologies.  It's _extemely_ fast, async, and very flexible in what it offers.  Furthermore, it's commercially supported and regularly updated.

It may bring TShock the following benefits:

* Sub-ms round-trip times
* The full power of the Web API 2.0 routing engine, including attribute routes, templates, etc
* Fully async
* Potential other OWIN features, such as SSO OAuth support (think steam authentication facebook, et. al)
* CORS support
* Possibility of a web interface via controllers and Razor views
* Pretty much anything else in the web that runs on WebAPI 2.0
* Request/respond via any serialization type: JSON, XML, Plain text without writing any extra code

The web API 2.0 stack is available from NuGet, and is embeddable into any application. 

More information here: http://www.asp.net/web-api/overview/hosting-aspnet-web-api/use-owin-to-self-host-web-api

If a move to a full supported OWIN stack is not ideal, the replacement REST interface should at least be faster.

**Comment by Ijwu** [comment]
I like the idea. It would be beneficial to directly port the current REST API over to use OWIN so that we don't break compatibility. All REST API v4 additions noted in #867 can be included here as well.

Pinging @nicatronTg: is this something we can work on to include in 5.0 or would you prefer to copy-pasta our current REST model to Orion for the sake of sticking with the feature freeze?

I think this is doable in 5.0 and would be nice to have.

**Comment by nicatronTg** [comment]
I think that it's entirely reasonable to use this rather than our own homebrew HttpServer. @Ijwu @WhiteXZ might have something better to say on this per Orion, but I think that it's definitely a go-ahead in Orion. If White's goal is to ship Orion with 1.3, then that means that it needs to be ready, implemented, and working by the 30th. If the goal is to have Orion done at some indefinite point in time in the future, then that's entirely out the window.

In short, we definitely need this, and definitely sooner, rather than later. We need to discuss what we want done by 1.3, and what we want by "some date in the future." We should also consider doing a soft-conversion to Orion and just using it now, rather than doing a giant switchover in the future.

**Comment by nicatronTg** [comment]
_The project lead for Orion is really @WhiteXZ_

**Comment by WhiteXZ** [comment]
I highly doubt Orion will be ready by the 30th :(
I am fully in favour of this improvement though

**Comment by tlworks** [comment]
Don't you think it's time for an improvement then? Considering it's almost a year further and nothing really has changed. Just my personal opinion.

**Comment by WhiteXZ** [comment]
Orion will be using OTAPI's OWIN capabilities.
